src/bridge/SequencerInbox.sol

```diff
@@ -356,24 +490,136 @@ contract SequencerInbox is DelegateCallAware, GasRefundEnabled, ISequencerInbox
-        return (keccak256(header), timeBounds);
+        uint256 fullDataLen = HEADER_LENGTH + data.length;
+        if (fullDataLen > maxDataSize) revert DataTooLarge(fullDataLen, maxDataSize);
+
+        (bytes memory header, IBridge.TimeBounds memory timeBounds) = packHeader(
+            afterDelayedMessagesRead
+        );
+
+        // the batch poster is allowed to submit an empty batch, they can use this to progress the
+        // delayed inbox without providing extra batch data
+        if (data.length > 0) {
+            // The first data byte cannot be the same as any that have been set via other methods (eg 4844 blob header) as this
+            // would allow the supplier of the data to spoof an incorrect 4844 data batch
+            if (!isValidCallDataFlag(data[0])) revert InvalidHeaderFlag(data[0]);
+
+            // the first byte is used to identify the type of batch data
+            // das batches expect to have the type byte set, followed by the keyset (so they should have at least 33 bytes)
+            // if invalid data is supplied here the state transition function will process it as an empty block
+            // however we can provide a nice additional check here for the batch poster
+            if (data[0] & DAS_MESSAGE_HEADER_FLAG != 0 && data.length >= 33) {
+                // we skip the first byte, then read the next 32 bytes for the keyset
+                bytes32 dasKeysetHash = bytes32(data[1:33]);
+                if (!dasKeySetInfo[dasKeysetHash].isValidKeyset) revert NoSuchKeyset(dasKeysetHash);
+            }
+        }
+        return (keccak256(bytes.concat(header, data)), timeBounds);
+    }
+
+    /// @dev   Form a hash of the data being provided in 4844 data blobs
+    /// @param afterDelayedMessagesRead The delayed messages count read up to
+    /// @return The data hash
+    /// @return The timebounds within which the message should be processed
+    function formBlobDataHash(uint256 afterDelayedMessagesRead)
+        internal
+        view
+        returns (bytes32, IBridge.TimeBounds memory)
+    {
+        bytes32[] memory dataHashes = dataHashReader.getDataHashes();
+        if (dataHashes.length == 0) revert MissingDataHashes();
+
+        (bytes memory header, IBridge.TimeBounds memory timeBounds) = packHeader(
+            afterDelayedMessagesRead
+        );
+
+        return (
+            keccak256(bytes.concat(header, DATA_BLOB_HEADER_FLAG, abi.encodePacked(dataHashes))),
+            timeBounds
+        );
+    }
+
+    /// @dev   Submit a batch spending report message so that the batch poster can be reimbursed on the rollup
+    /// @param dataHash The hash of the message the spending report is being submitted for
+    /// @param seqMessageIndex The index of the message to submit the spending report for
+    /// @param gasPrice The gas price that was paid for the data (standard gas or data gas)
+    function submitBatchSpendingReport(
```

Comment:
This is now almost duplicate code of what we have in addSequencerL2BatchImpl, should we call this instead? Or are you wanting to do that refactor in a later PR? if so we should leave a comment addSequencerL2BatchImpl explaining why we havent done it now

Reply:
This would be a breaking change if we use tx.origin, and they are not getting the refund immediately anyway (it would be on L2) so I rather to have to current behavior.